FEATURE: get rid of the import a query modal

AndrewPrigorshnev opened the pull request:
This is the new version of https://github.com/discourse/discourse-data-explorer/pull/126. Now the `pick-files-button` moved to core. 

To import a query into Data Explorer, you need to push the button and then deal with the import modal:

![124735139-d168f500-df26-11eb-8b71-82f0ada18129](https://user-images.githubusercontent.com/1274517/125976211-4a48420f-74b1-46a4-84da-bd8826174f58.png)

![124735161-d75ed600-df26-11eb-943b-024c873df6da](https://user-images.githubusercontent.com/1274517/125976226-8385dfc6-22f2-4933-b3ff-f967b7c3dc22.png)

Instead, we want just to be triggering a system file picker directly. This PR makes it happen.

martin-brennan commented on assets/javascripts/discourse/controllers/admin-plugins-explorer.js.es6:
```diff
@@ -30,6 +30,16 @@ export default Ember.Controller.extend({
 
+  @computed
+  acceptedImportFileTypes() {
+    return [
```
I am just wondering why we need all these variants of the JSON type? Do different browsers send different things? Usually it is application/json

AndrewPrigorshnev replied:
I don't know initial intentions behind this list, I've just moved it from other place in this PR. But looks like you're right and it's more than enough to have just `application/json`. For example, [text/javascript is obsolete, and application/x-javascript was experimental ](https://stackoverflow.com/a/9664327/3206146).

AndrewPrigorshnev replied:
And [the mime type for json text is application/json](https://stackoverflow.com/a/477819/3206146).

We also already use just `application/json` with file inputs in several places, it seems to work fine.

martin-brennan commented on assets/javascripts/discourse/controllers/admin-plugins-explorer.js.es6:
```diff
@@ -117,16 +127,41 @@ export default Ember.Controller.extend({
+  },
+
+  _importQuery(file) {
+    const query = JSON.parse(file).query;
```
I think we should add some JSON parse error handling here, in case the JSON in the file is not correctly structured. Then you can show a message to the user and reset the `loading` state

martin-brennan replied:
Also do we know what the structure should look like ahead of time? Maybe we should check for the keys that are required and throw an error if the user doesn't provide them?

martin-brennan commented on assets/javascripts/discourse/controllers/admin-plugins-explorer.js.es6:
```diff
@@ -117,16 +127,41 @@ export default Ember.Controller.extend({
 
+  _readQueryFile(file) {
+    const reader = new FileReader();
+    reader.onload = (evt) => {
```
Probably want to implement `FileReader.onerror` here as well to be able to reset the state and show a message on error. If this is a pattern we use in a few places in future we may want to refactor this out into a `lib` that we can import. No need for that yet though as this is the only place we are doing this so far :)

martin-brennan replied:
E.g. `app/lib/file-reader.js`

martin-brennan commented on assets/javascripts/discourse/controllers/admin-plugins-explorer.js.es6:
```diff
@@ -117,16 +127,41 @@ export default Ember.Controller.extend({
+        this.set("loading", false);
+        this.addCreatedRecord(q.target);
+      })
+      .catch(popupAjaxError);
```
You will need to set `loading` to false here too I assume?

martin-brennan reviewed (changes requested):
Looking good and I like the new component, just a few questions.

AndrewPrigorshnev commented:
All comments about error handling and setting `loading` to false are slightly out of the scope of this PR, I wanted just to get rid of an unnecessary modal and carefully move existing code. 

At the same time, they all are perfect catches, and we definitely need to fix them. So I fixed these problems and I also switched these methods to `async/await` pattern, all these checks implemented with promise chaining would look overcomplicated.

Speaking of extracting code for reading files to `lib`, this wrapper that can be awaited can be our starting point:
```javascript
  _readFileAsTextAsync(file) {
    return new Promise((resolve, reject) => {
      const reader = new FileReader();
      reader.onload = () => {
        resolve(reader.result);
      };
      reader.onerror = reject;

      reader.readAsText(file);
    });
  }
```

martin-brennan commented on assets/javascripts/discourse/controllers/admin-plugins-explorer.js.es6:
```diff
@@ -117,16 +124,64 @@ export default Ember.Controller.extend({
+    return query;
+  },
+
+  _readFileAsTextAsync(file) {
```
Shouldn't this function be marked `async` if you are `await`ing it in `_importQuery`?

AndrewPrigorshnev replied:
No, we don't need it here. Normally, you need `async` if you're awaiting something _inside_ the function. This function just create and return a promise, and we're awaiting that promise outside.

martin-brennan replied:
Ah cool I didn't realise that thanks for clarifying

martin-brennan commented on assets/javascripts/discourse/controllers/admin-plugins-explorer.js.es6:
```diff
@@ -117,16 +124,64 @@ export default Ember.Controller.extend({
+    return response.target;
+  },
+
+  async _parseQuery(json) {
```
I don't think this function needs to be `async`, nothing that is promise-related is happening here.

AndrewPrigorshnev replied:
Oh, this is redundant here, a good catch!

martin-brennan commented on assets/javascripts/discourse/controllers/admin-plugins-explorer.js.es6:
```diff
@@ -117,16 +124,64 @@ export default Ember.Controller.extend({
+      const file = files[0];
+      this._importQuery(file)
+        .then((record) => this.addCreatedRecord(record))
+        .catch((e) => {
```
Much better, thanks for adding the error handling

martin-brennan commented on config/locales/client.en.yml:
```diff
@@ -15,6 +15,8 @@ en:
         label: "Import"
         modal: "Import A Query"
+        unparseable_json: "Unparseable json file."
+        wrong_json: "Wrong json file. A json file should contain an 'query' object, which should at least has a 'sql' property."
```
Nitpick, should be `"which should at least have an 'sql' property"`. Also...I don't mind if you change this or not, it's just another nitpick, but I think JSON should be capitalised because it is an acronym

AndrewPrigorshnev replied:
I've changed both, thank you 👍, please feel free to comment such mistakes.

martin-brennan commented on assets/javascripts/discourse/controllers/admin-plugins-explorer.js.es6:
```diff
@@ -30,6 +32,11 @@ export default Ember.Controller.extend({
 
+  @computed
+  acceptedImportFileTypes() {
+    return ["application/json"];
```
Thanks for changing this. I think the other modal was done in 2014 so maybe this wasn't the accepted mime type back then? Anyway application/json is totally fine, never used anything but that.

martin-brennan reviewed (changes requested):
Much better, still a few minor changes though.

martin-brennan reviewed (approved):
LGTM :+1:
